File: R/logitnorm.R
#' Logit Normal Distribution
#'
#' Density, cumulative distribution, quantile functions and random number
#' generation for the distribution that becomes normal after the Logit
#' transformation.
#'
#' The distribution has the following density function:
#'
#' f(y) = 1/(sqrt(2 pi) y (1-y)) exp(-(logit(y) -mu)^2 / (2 sigma^2))
#'
#' where y is in (0, 1) and logit(y) =log(y/(1-y)).
#'
#' Both \code{plogitnorm} and \code{qlogitnorm} are returned for the lower
#' tail of the distribution.
#'
#' All the functions are defined for the values between 0 and 1.
#'
#' @template author
#' @keywords distribution
#'
#' @param q vector of quantiles.
#' @param p vector of probabilities.
#' @param n number of observations. Should be a single number.
#' @param mu vector of location parameters (means).
#' @param sigma vector of scale parameters.
#' @param log if \code{TRUE}, then probabilities are returned in
#' logarithms.
#'
#' @return Depending on the function, various things are returned
#' (usually either vector or scalar):
#' \itemize{
#' \item \code{dlogitnorm} returns the density function value for the
#' provided parameters.
#' \item \code{plogitnorm} returns the value of the cumulative function
#' for the provided parameters.
#' \item \code{qlogitnorm} returns quantiles of the distribution. Depending
#' on what was provided in \code{p}, \code{mu} and \code{sigma}, this
#' can be either a vector or a matrix, or an array.
#' \item \code{rlogitnorm} returns a vector of random variables
#' generated from the logitnorm distribution. Depending on what was
#' provided in \code{mu} and \code{sigma}, this can be either a vector
#' or a matrix or an array.
#' }
#'
#' @examples
#' x <- dlogitnorm(c(-1000:1000)/200, 0, 1)
#' plot(c(-1000:1000)/200, x, type="l")
#'
#' x <- plogitnorm(c(-1000:1000)/200, 0, 1)
#' plot(c(-1000:1000)/200, x, type="l")
#'
#' qlogitnorm(c(0.025,0.975), 0, c(1,2))
#'
#' x <- rlogitnorm(1000, 0, 1)
#' hist(x)
#'
#' @references \itemize{
#' \item Mead, R. (1965). A Generalised Logit-Normal Distribution.
#' Biometrics, 21 (3), 721–732. doi: 10.2307/2528553
#' }
#'
#' @rdname logitnorm-distribution
#' @importFrom stats dnorm pnorm qnorm rnorm
#' @export dlogitnorm
dlogitnorm <- function(q, mu=0, sigma=1, log=FALSE){
    logitnormReturn <- 1 / (sigma*sqrt(2*pi)*q*(1-q))*exp(-(log(q/(1-q))-mu)^2/(2*sigma^2));
    # logitnormReturn[q<=0] <- -Inf;
    # logitnormReturn[q>=1] <- Inf;

    # Return logs if needed
    if(log){
        logitnormReturn[] <- log(logitnormReturn);
    }

    return(logitnormReturn);
}

#' @rdname logitnorm-distribution
#' @export plogitnorm
#' @aliases plogitnorm
plogitnorm <- function(q, mu=0, sigma=1){
    logitnormReturn <- vector("numeric", length(q));
    logitnormReturn[q>=0] <- pnorm(q=log(q/(1-q)), mean=mu, sd=sigma);
    logitnormReturn[q<0] <- 0;
    logitnormReturn[q>=1] <- 1;
    return(logitnormReturn);
}

#' @rdname logitnorm-distribution
#' @export qlogitnorm
#' @aliases qlogitnorm
qlogitnorm <- function(p, mu=0, sigma=1){
    logitnormReturn <- qnorm(p=p, mean=mu, sd=sigma);
    logitnormReturn[] <- exp(logitnormReturn)/(1+exp(logitnormReturn));
    logitnormReturn[is.nan(logitnormReturn)] <- 0;
    return(logitnormReturn);
}

#' @rdname logitnorm-distribution
#' @export rlogitnorm
#' @aliases rlogitnorm
rlogitnorm <- function(n=1, mu=0, sigma=1){
    logitnormReturn <- qlogitnorm(runif(n=n, 0, 1), mu=mu, sigma=sigma);
    return(logitnormReturn);
}

File: python/src/greybox/distributions/logitnorm.py
"""Logit-Normal distribution functions.

Density, cumulative distribution, quantile functions and random number
generation for the distribution that becomes normal after the Logit
transformation.
"""

import numpy as np
from scipy import stats


def dlogitnorm(q, loc=0, scale=1, log=False):
    """Logit-Normal distribution density.

    f(y) = 1/(sqrt(2*pi)*scale*y*(1-y)) * exp(-(logit(y) - loc)^2 / (2*scale^2))

    Parameters
    ----------
    q : array_like
        Quantiles (must be in (0, 1)).
    loc : float
        Location parameter (on logit scale).
    scale : float
        Scale parameter.
    log : bool
        If True, return log-density.

    Returns
    -------
    array
        Density values.
    """
    q = np.clip(q, 1e-10, 1 - 1e-10)
    logit_q = np.log(q / (1 - q))
    density = (
        1
        / (scale * np.sqrt(2 * np.pi) * q * (1 - q))
        * np.exp(-((logit_q - loc) ** 2) / (2 * scale**2))
    )
    density = np.maximum(density, 1e-300)
    if log:
        return np.log(density)
    return density


def plogitnorm(q, loc=0, scale=1):
    """Logit-Normal distribution CDF.

    Parameters
    ----------
    q : array_like
        Quantiles.
    loc : float
        Location parameter.
    scale : float
        Scale parameter.

    Returns
    -------
    array
        CDF values.
    """
    q = np.asarray(q)
    result = np.zeros_like(q, dtype=float)

    mask_negative = q < 0
    mask_between = (q >= 0) & (q <= 1)
    mask_above = q > 1

    result[mask_negative] = 0
    result[mask_above] = 1

    if np.any(mask_between):
        q_between = q[mask_between]
        result[mask_between] = stats.norm.cdf(
            np.log(q_between / (1 - q_between)), loc=loc, scale=scale
        )

    return result


def qlogitnorm(p, loc=0, scale=1):
    """Logit-Normal distribution quantile function.

    Parameters
    ----------
    p : array_like
        Probabilities.
    loc : float
        Location parameter.
    scale : float
        Scale parameter.

    Returns
    -------
    array
        Quantile values.
    """
    p = np.asarray(p)
    result = np.exp(stats.norm.ppf(p, loc=loc, scale=scale)) / (
        1 + np.exp(stats.norm.ppf(p, loc=loc, scale=scale))
    )
    result = np.where(np.isnan(result), 0, result)
    return result


def rlogitnorm(n, loc=0, scale=1):
    """Logit-Normal distribution random number generation.

    Parameters
    ----------
    n : int
        Number of observations.
    loc : float
        Location parameter.
    scale : float
        Scale parameter.

    Returns
    -------
    array
        Random values.
    """
    return qlogitnorm(np.random.uniform(0, 1, n), loc, scale)
